README.md
# rc-date-core ![npm version](https://img.shields.io/npm/v/rc-date-core.svg?style=flat)
A compact and delicate crafted core of date input.

<img src="http://7xib2u.com1.z0.glb.clouddn.com/daypicker.png" width="236"/>

## Installation
``` shell
npm install rc-date-core --save
```

## Props
### date picker props
- **onChange** (func) return string value formatted by prop 'mode' and 'returnFormat' and date object
- **value** (string, date object, int) date value of input e.g. 'yyyy/MM/dd' , 1464081349000
- **min** (string) min date limit depended on mode e.g. '2015/01/01'
- **max** (string) max date limit same as above
- **mode** (string default 'day')  'year' 'month' 'day' 'hour' 'minute' 'second'
- **returnFormat** (string) e.g. 'yyyy/MM/dd' if none provided, just decided by 'mode'
- **className** (string)

### date input props
- ...any props above
- **autoPosition** (bool default false) enable picker changing position when scrolling
- **preferPosition** (string default "bottomLeft") initial position of picker(als0 will affect autoPosition when enabled)
- **container** (selector or element default window) scroll container for autoPosition
- **closeOnSelect** (bool default true) close picker when select
- **closeOnClickOutside** (bool default true) close picker when click outside
- **displayFormat** (string) e.g. 'yyyy/MM/dd' input value display, if none provided, just decided by 'mode'
- ...any other props of react input element e.g. 'readOnly' 'disabled'


## Usage
```es6
import {DatePicker} from 'rc-date-core'
// use dark skin
import Styles from './node_modules/rc-date-core/dist/darkInput.css'
class Demo extends React.Component{
    constructor(){
        this.state = {
            value: '2016/03/01'
            inputValue: '2016/03/02'
        }
    }
    render(){
        return (
            <div>
                <DatePicker ref="datepicker"
                       mode="day"
                       min="2016/01/01"
                       max="2016/11/01"
                       value={this.state.value}
                       onChange={(v)=>{this.setState({value: v})}}
               />
               <DatePickerInput
                    mode="day"
                    min="2016/01/01"
                    max="2016/11/01"
                    value={this.state.inputValue}
                    readOnly={true}
                    autoPosition={true}
                    container={window}
                    preferPosition="bottomRight"
                    closeOnSelect={true}
                    closeOnClickOutside={true}
                    onChange={(v)=>{this.setState({inputValue: v})}}
                />
            </div>
       )
    }
}
```


HISTORY.md
1.2.1 / 2016-08-09
==================
 * FIX: make each view can only change its field value
 
1.2.0 / 2016-08-09
==================
 * FEATURE: support time picker
 
1.1.7 / 2016-05-24
==================
 * FEATURE: support more value types
 


index.js
import React, {PropTypes} from 'react'
import ReactDOM from 'react-dom'

import {DatePicker, DatePickerInput} from './src'
import Styles from './dist/darkInput.css'
import DemoStyles from './index.css'

class Page extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            valueScroll: '2016-03-14',
            value: new Date(),
            valueNumber: Date.now(),
            valueTime: '2016-03-14 12:12',
            valueFullDay: '2016-03-14 12:12:12',
            valueDay: '2016-03-14',
            valueMonth: '2016-03',
            valueYear: new Date(),
        };
    }

    render() {
        return <div className="demos">
            <h3>Date input(auto position in scroll container)</h3>
            <div id="scrollContainer" className="wrapper-limited">
                <div className="section">
                <div className="col col1">
                    <DatePickerInput
                        value={this.state.valueScroll}
                        readOnly={true}
                        autoPosition={true}
                        container='#scrollContainer'
                        preferPosition="topLeft"
                        closeOnSelect={true}
                        closeOnClickOutside={true}
                        displayFormat="yyyy/MM/dd"
                        returnFormat="yyyy-MM-dd"
                        onChange={(v)=>{this.setState({valueScroll: v})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueScroll}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueScroll: '2020-02-02'})
                    }}>Set date to 2020-02-02</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueScroll: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        value= <span className="pl-s"><span className="pl-pds">"</span>2016-03-14<span className="pl-pds">"</span></span><br/>
                        autoPosition= <span className="pl-k">true</span><br/>
                        container= <span className="pl-k">'#scrollContainer'</span><br/>
                        preferPosition= <span className="pl-s"><span className="pl-pds">"</span>topLeft<span className="pl-pds">"</span></span><br/>
                        closeOnSelect= <span className="pl-k">true</span><br/>
                        closeOnClickOutside= <span className="pl-k">true</span><br/>
                        displayFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy/MM/dd<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>
            </div>

            <h3>Date input(auto position)</h3>
            <div className="section">
                <div className="col col1">
                    <DatePickerInput
                        value={this.state.value}
                        readOnly={true}
                        autoPosition={true}
                        preferPosition="bottomRight"
                        closeOnSelect={true}
                        closeOnClickOutside={true}
                        displayFormat="yyyy年M月d日"
                        returnFormat="yyyy-MM-dd"
                        onChange={(v, date)=>{this.setState({value: date})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.value && this.state.value.toString()}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({value: '2020-02-02'})
                    }}>Set date to 2020-02-02</span>
                        <span className="btn" onClick={()=>{
                        this.setState({value: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        value= <span className="pl-k">new </span>Date()<br/>
                        autoPosition= <span className="pl-k">true</span><br/>
                        preferPosition= <span className="pl-s"><span className="pl-pds">"</span>bottomRight<span className="pl-pds">"</span></span><br/>
                        closeOnSelect= <span className="pl-k">true</span><br/>
                        closeOnClickOutside= <span className="pl-k">true</span><br/>
                        displayFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy年M月d日<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Date input(different type of value)</h3>
            <div className="section">
                <div className="col col1">
                    <DatePickerInput
                        value={this.state.valueNumber}
                        readOnly={true}
                        autoPosition={true}
                        preferPosition="bottomRight"
                        closeOnSelect={true}
                        closeOnClickOutside={true}
                        displayFormat="yyyy年M月d日 HH:mm"
                        returnFormat="yyyy-MM-dd"
                        onChange={(v, date)=>{this.setState({valueNumber: date && date.getTime()})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueNumber && this.state.valueNumber.toString()}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueNumber: '2020-02-02'})
                    }}>Set date to 2020-02-02</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueNumber: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        value= Date.now()<br/>
                        autoPosition= <span className="pl-k">true</span><br/>
                        preferPosition= <span className="pl-s"><span className="pl-pds">"</span>bottomRight<span className="pl-pds">"</span></span><br/>
                        closeOnSelect= <span className="pl-k">true</span><br/>
                        closeOnClickOutside= <span className="pl-k">true</span><br/>
                        displayFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy年M月d日<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Date input(time picker)</h3>
            <div className="section">
                <div className="col col1">
                    <DatePickerInput
                        value={this.state.valueTime}
                        readOnly={true}
                        autoPosition={true}
                        preferPosition="bottomRight"
                        closeOnSelect={false}
                        closeOnClickOutside={true}
                        mode="minute"
                        displayFormat="yyyy年M月d日 HH:mm"
                        returnFormat="yyyy-MM-dd HH:mm"
                        onChange={(v, date)=>{this.setState({valueTime: v})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueTime && this.state.valueTime.toString()}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueNumber: '2020-02-02 12:12'})
                    }}>Set date to 2020-02-02 12:12</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueTime: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        value= Date.now()<br/>
                        autoPosition= <span className="pl-k">true</span><br/>
                        preferPosition= <span className="pl-s"><span className="pl-pds">"</span>bottomRight<span className="pl-pds">"</span></span><br/>
                        closeOnSelect= <span className="pl-k">false</span><br/>
                        closeOnClickOutside= <span className="pl-k">true</span><br/>
                        mode= <span className="pl-s"><span className="pl-pds">"</span>minute<span className="pl-pds">"</span></span><br/>
                        displayFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy年M月d日 HH:mm<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd HH:mm<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Day picker</h3>
            <div className="section">
                <div className="col col1">
                    <DatePicker
                        mode="day"
                        min="2015-05-20"
                        max="2030-05-20"
                        value={this.state.valueDay}
                        returnFormat="yyyy-MM-dd"
                        onChange={(v)=>{this.setState({valueDay: v})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueDay}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueDay: '2020-02-02'})
                    }}>Set date to 2020-02-02</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueDay: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        mode= <span className="pl-s"><span className="pl-pds">"</span>day<span className="pl-pds">"</span></span><br/>
                        min= <span className="pl-s"><span className="pl-pds">"</span>2015-05-20<span className="pl-pds">"</span></span><br/>
                        max= <span className="pl-s"><span className="pl-pds">"</span>2030-05-20<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Day picker(With hour&minute)</h3>
            <div className="section">
                <div className="col col1">
                    <DatePicker
                        mode="second"
                        min="2015-05-20"
                        max="2030-05-20"
                        value={this.state.valueFullDay}
                        returnFormat="yyyy-MM-dd HH:mm:ss"
                        onChange={(v)=>{this.setState({valueFullDay: v})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueFullDay}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueFullDay: '2020-02-02 12:12:12'})
                    }}>Set date to 2020-02-02 12:12:12</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueFullDay: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        mode= <span className="pl-s"><span className="pl-pds">"</span>second<span className="pl-pds">"</span></span><br/>
                        min= <span className="pl-s"><span className="pl-pds">"</span>2015-05-20<span className="pl-pds">"</span></span><br/>
                        max= <span className="pl-s"><span className="pl-pds">"</span>2030-05-20<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM-dd HH:mm:ss<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Month picker</h3>
            <div className="section">
                <div className="col col1">
                    <DatePicker
                        mode="month"
                        min="2015-05"
                        max="2030-05"
                        returnFormat="yyyy-MM"
                        value={this.state.valueMonth}
                        onChange={(v)=>{this.setState({valueMonth: v})}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueMonth}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueMonth: '2020-02'})
                    }}>Set date to 2020-02</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueMonth: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        mode= <span className="pl-s"><span className="pl-pds">"</span>month<span className="pl-pds">"</span></span><br/>
                        min= <span className="pl-s"><span className="pl-pds">"</span>2015-05<span className="pl-pds">"</span></span><br/>
                        max= <span className="pl-s"><span className="pl-pds">"</span>2030-05<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy-MM<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>

            <h3>Year picker</h3>
            <div className="section">
                <div className="col col1">
                    <DatePicker
                        mode="year"
                        min="2015"
                        max="2030"
                        returnFormat="yyyy"
                        value={this.state.valueYear}
                        onChange={(v, date)=>{this.setState({valueYear: date});}}
                    />
                </div>
                <div className="col col2">
                    <p>Selected: <span className="value">{this.state.valueYear && this.state.valueYear.toString()}</span></p>
                    <p className="btn-group">
                        <span className="btn" onClick={()=>{
                        this.setState({valueYear: '2020-02'})
                    }}>Set date to 2020</span>
                        <span className="btn" onClick={()=>{
                        this.setState({valueYear: null})
                    }}>Clear</span>
                    </p>
                    <p>Props</p>
                    <pre>
                        mode= <span className="pl-s"><span className="pl-pds">"</span>year<span className="pl-pds">"</span></span><br/>
                        min= <span className="pl-s"><span className="pl-pds">"</span>2015<span className="pl-pds">"</span></span><br/>
                        max= <span className="pl-s"><span className="pl-pds">"</span>2030<span className="pl-pds">"</span></span><br/>
                        returnFormat= <span className="pl-s"><span className="pl-pds">"</span>yyyy<span className="pl-pds">"</span></span><br/>
                    </pre>
                </div>
            </div>
        </div>
    }
}

ReactDOM.render(<Page />, document.getElementById('main'));
